homeassistant/components/lock/__init__.py

```diff
@@ -87,40 +87,34 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     return True
 
 
-async def _async_lock(entity: LockEntity, service_call: ServiceCall) -> None:
-    """Lock the lock."""
+@callback
+def _add_default_code(entity: LockEntity, service_call: ServiceCall) -> dict[Any, Any]:
     code: str = service_call.data.get(
         ATTR_CODE, entity._lock_option_default_code  # pylint: disable=protected-access
     )
     if entity.code_format_cmp and not entity.code_format_cmp.match(code):
         raise ValueError(
             f"Code '{code}' for locking {entity.entity_id} doesn't match pattern {entity.code_format}"
         )
-    await entity.async_lock(**remove_entity_service_fields(service_call))
+    data = remove_entity_service_fields(service_call)
+    if code:
+        data[ATTR_CODE] = code
```

Comment:
code is already in so maybe `if not code:`?

Reply:
`code` is optional in the service schema, right?

Reply:
It is but if I remember right an empty code from frontend is passed as `""` and not `None`.
So if `code` is `""` it would not replace it here and a service call would be made with it, right?

Maybe I'm just confusing myself on this one but just want to be sure we handle if it's `None`, `""` or anything legit (with or without the format check)

Reply:
If a service call is made with a code which is `""` (frontend would do that unless fixed since I looked at it) and we don't have a default code it will continue down to the integration with the code of `""`.
I think that would be wrong and we need to handle that.

Reply:
Then we should pop the item from the service data at the top of the body here.

Reply:
Yes, there seems to be something broken with the frontend lock UI, but I'm not sure what it is.

Locking/unlocking by pressing the text "LOCK"/"UNLOCK" (to the right) sends a service call without any code, and works with the fix in this PR.
<img width="499" alt="image" src="https://github.com/home-assistant/core/assets/5984518/71f3113f-4d9f-4b97-a8b8-b3bfeec54489">

Locking/unlocking from the entity view without entering a code sends an error message:
<img width="595" alt="image" src="https://github.com/home-assistant/core/assets/5984518/de0e5851-8ec6-45b8-a8e9-92c9119f7835">
```
{"type":"call_service","domain":"system_log","service":"write","service_data":{"logger":"frontend.js.latest.202310302","message":"Unhandled promise rejection from Chrome 119.0.0.0 on Mac OS 10.15.7\nError: Code dialog closed\nhass.localize (src/data/lock.ts:42:12)","level":"debug"},"id":64}
```

I will make a separate bug report on the frontend this this.

I am still happy to remove an empty code from the service call, but that will not fix the UI…

Reply:
@gjohansson-ST if the frontend passes an empty string, the existing code is broken in any case:
```
    code: str = service_call.data.get(
        ATTR_CODE, entity._lock_option_default_code  # pylint: disable=protected-access
    )
```
Are we sure we want to use the default code if an empty string is passed explicitly?

Reply:
Yes. Also @MartinHjelmare gave the same comment before. An empty string should not be considered a valid code.
I can try to take a look at frontend latest tomorrow and see what's going on. I guess it's limited to the lock card.

Reply:
OK, updated to do the following:

1. If `code` is passed in as `None` or an empty string, we use the default code.
2. If code format is set, we validate the code.
3. If `code` still is `None` or an empty  string, it is removed before forwarding the call.

Since `service_call.data` is a read only dictionary, I thought it made most sense to wait with popping `code` until after it is converted to a regular dictionary with `remove_entity_service_fields()`

Reply:
Simplified the code by always popping the code.

Reply:
I gave the code a go (along with trying to check frontend and it works as expected while the frontend issue remains).
Will check the tests etc. later but we should be able to finalize this pretty shortly and try to get a frontend fix.

Reply:
I changed the title but can you fix the PR description @oyvindwe 👍

Reply:
PR description updated!

Reply:
@gjohansson-ST I think you need to approve the PR as well, as you requested changes.